Multiple target patterns error for CMake

Java -version output
--------------------
$ java -version
openjdk version "11.0.22" 2024-01-16
IBM Semeru Runtime Open Edition 11.0.22.0 (build 11.0.22+7)
Eclipse OpenJ9 VM 11.0.22.0 (build openj9-0.43.0, JRE 11 Windows Server 2019 amd64-64-Bit Compressed References 20240131_835 (JIT enabled, AOT enabled)
OpenJ9   - 2c3d78b48
OMR      - ea8124dbc


Summary of problem
------------------
While running build compile for a windows 19 machine (with VS2022 ) I am running into `multiple target patterns.  Stop. make[6]: *** [CMakeFiles/Makefile2:8528: runtime/compiler/CMakeFiles/j9jit.dir/all] Error 2` error while running make all.

I have faced this issue before and by rolling down version of cmake to v3.20.1 helped me avoid this error. Since cygwin installer only has recent versions of packages to be installed for the recent cygwin installer cmake v3.20.1 is unavailable, only v3.23 and higher is available at the moment.

job reference:
https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK11_x86-64_windows_Personal/1595/console





Diagnostic files
----------------
```
[2024-04-27T15:13:55.413Z] [ 51%] Built target j9gcbase
[2024-04-27T15:14:01.747Z] Consolidate compiler generated dependencies of target j9jit
[2024-04-27T15:14:03.595Z] runtime/compiler/CMakeFiles/j9jit.dir/compiler_depend.make:287797: *** multiple target patterns.  Stop.
[2024-04-27T15:14:03.595Z] make[6]: *** [CMakeFiles/Makefile2:8528: runtime/compiler/CMakeFiles/j9jit.dir/all] Error 2
[2024-04-27T15:14:03.595Z] make[6]: *** Waiting for unfinished jobs....
[2024-04-27T15:14:06.083Z] Consolidate compiler generated dependencies of target omrgc
[2024-04-27T15:14:07.364Z] [ 64%] Built target omrgc
[2024-04-27T15:14:07.364Z] make[5]: *** [CMakeFiles/Makefile2:3351: runtime/CMakeFiles/j9ddr.dir/rule] Error 2
[2024-04-27T15:14:07.364Z] make[4]: *** [Makefile:238: runtime/CMakeFiles/j9ddr.dir/rule] Error 2
[2024-04-27T15:14:07.364Z] make[3]: *** [/cygdrive/c/users/jenkins/workspace/build_jdk11_x86-64_windows_personal/closed/OpenJ9.gmk:566: build-j9] Error 2
[2024-04-27T15:14:07.364Z] make[2]: *** [/cygdrive/c/users/jenkins/workspace/build_jdk11_x86-64_windows_personal/closed/custom/Main.gmk:50: j9vm-build] Error 2
[2024-04-27T15:14:07.364Z] 
[2024-04-27T15:14:07.364Z] ERROR: Build failed for target 'all' in configuration 'windows-x86_64-normal-server-release' (exit code 2) 
[2024-04-27T15:14:10.618Z] 
[2024-04-27T15:14:10.618Z] No indication of failed target found.
```

I tried uninstalling cmake from cygwin and installed v3.20.1 from source and linked cygwins /usr/bin/cmake to the installed directory. But for that the build is using Nmake instead of cmake .

```
21:02:03  ====
21:02:03  -- Building for: NMake Makefiles
21:02:04  -- The C compiler identification is unknown
21:02:05  -- The CXX compiler identification is unknown
21:02:05  CMake Error at CMakeLists.txt:25 (project):
21:02:05    The CMAKE_C_COMPILER:
21:02:05  
21:02:05      /cygdrive/c/users/jenkins/workspace/build_jdk11_x86-64_windows_personal/build/windows-x86_64-normal-server-release/tools/cl
21:02:05  
21:02:05    is not a full path to an existing compiler tool.
21:02:05  
21:02:05    To use the NMake generator with Visual C++, cmake must be run from a shell
21:02:05    that can use the compiler cl from the command line.  This environment is
21:02:05    unable to invoke the cl compiler.  To fix this problem, run cmake from the
21:02:05    Visual Studio Command Prompt (vcvarsall.bat).
21:02:05  
21:02:05    Tell CMake where to find the compiler by setting either the environment
21:02:05    variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
21:02:05    the compiler, or to the compiler name if it is in the PATH.
21:02:05  
21:02:05  
21:02:05  CMake Error at CMakeLists.txt:25 (project):
21:02:05    The CMAKE_CXX_COMPILER:
21:02:05  
21:02:05      /cygdrive/c/users/jenkins/workspace/build_jdk11_x86-64_windows_personal/build/windows-x86_64-normal-server-release/tools/cl
21:02:05  
21:02:05    is not a full path to an existing compiler tool.
21:02:05  
21:02:05    To use the NMake generator with Visual C++, cmake must be run from a shell
21:02:05    that can use the compiler cl from the command line.  This environment is
21:02:05    unable to invoke the cl compiler.  To fix this problem, run cmake from the
21:02:05    Visual Studio Command Prompt (vcvarsall.bat).
21:02:05  
21:02:05    Tell CMake where to find the compiler by setting either the environment
21:02:05    variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
21:02:05    to the compiler, or to the compiler name if it is in the PATH.
21:02:05  
21:02:05  
21:02:05  -- Configuring incomplete, errors occurred!
```

--- a/runtime/cmake/platform/os/win.cmake
+++ b/runtime/cmake/platform/os/win.cmake
@@ -20,34 +20,35 @@
 # SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
 ################################################################################
 
-# Note: we need to inject WIN32 et al, as OMR no longer uses them
+# Note: Define WIN32 etc. until OMR no longer uses them.
 
 list(APPEND OMR_PLATFORM_DEFINITIONS
-    -DWIN32
-    -D_WIN32
+	-DWIN32
+	-D_WIN32
 )
 if(OMR_ENV_DATA64)
-    list(APPEND OMR_PLATFORM_DEFINITIONS
-        -DWIN64
-        -D_WIN64
-    )
+	list(APPEND OMR_PLATFORM_DEFINITIONS
+		-DWIN64
+		-D_WIN64
+	)
 endif()
+
 list(APPEND OMR_PLATFORM_DEFINITIONS -DWINDOWS)
 
-# Set flags we use to build the interpreter
+# Set flags we use to build the interpreter.
 omr_stringify(CMAKE_J9VM_CXX_FLAGS
-    -O3
-    -fno-rtti
-    -fno-threadsafe-statics
-    -fno-strict-aliasing
-    -fno-exceptions
-    -fno-asynchronous-unwind-tables
-    -std=c++0x
-    -D_CRT_SUPPRESS_RESTRICT
-    -DVS12AndHigher
-    ${OMR_PLATFORM_DEFINITIONS}
+	-O3
+	-fno-rtti
+	-fno-threadsafe-statics
+	-fno-strict-aliasing
+	-fno-exceptions
+	-fno-asynchronous-unwind-tables
+	-std=c++0x
+	-D_CRT_SUPPRESS_RESTRICT
+	-DVS12AndHigher
+	${OMR_PLATFORM_DEFINITIONS}
 )
 
 if(OMR_ENV_DATA32)
-    set(CMAKE_J9VM_CXX_FLAGS "${CMAKE_J9VM_CXX_FLAGS} -m32")
+	set(CMAKE_J9VM_CXX_FLAGS "${CMAKE_J9VM_CXX_FLAGS} -m32")
 endif()

--- a/runtime/compiler/CMakeLists.txt
+++ b/runtime/compiler/CMakeLists.txt
@@ -38,6 +38,12 @@ if(OMR_ARCH_X86)
 		endif()
 	endif()
 	enable_language(ASM_NASM)
+	if(OMR_OS_WINDOWS)
+		# On Windows, automatic dependency tracking for ASM_NASM files doesn't play
+		# nicely with fixpath.sh. Remove the nasm options that would produce *.d
+		# files: cmake silently treats missing *.d files as if they were empty.
+		set(CMAKE_DEPFILE_FLAGS_ASM_NASM "")
+	endif()
 	# We have to manually append "/" to the paths as NASM versions older than v2.14 requires trailing / in the directory paths.
 	set(asm_inc_dirs
 		"-I${j9vm_SOURCE_DIR}/oti/"

